### PR TITLE
Support reading labels for paths from a YML file

### LINKED ISSUE #5: Support defining labels inside codeowners file
As a first iteration, let's add the labels on each path.

## Benefits:
- OSS
- Do not centralise labels management in the probot, delegating to each repo

## Follow-up
Define macros per team to reuse labels across, which adds extensibility to the codewoners file

- use @@PROBOT-ID to define an use probots
- a probot ID is separated from its values with `:`
- probot's values are separated by commas (`,`)
- probot's values are enclosed in double quotes
- use a macro
- extend macros per line
- define multiple probots per line

A possible regex for the protocol: `@@([[alphanum]]-_)+:(".+"|.+)`

```gitignore
# @@probot-codeowners-labeller/apm:"Team:apm","Team:logs-ui"
# @@probot-slack/apm:"#channel1","#channel2"
# APM
/x-pack/legacy/plugins/apm/  @elastic/apm-ui         # @@probot-codeowners-labeller/apm @@probot-slack/apm
/x-pack/plugins/apm/  @elastic/apm-ui
/x-pack/test/functional/apps/apm/  @elastic/apm-ui   # @@probot-codeowners-labeller/apm:"label2","label3","label4" @@probot-slack/apm:"@manu"
/src/legacy/core_plugins/apm_oss/  @elastic/apm-ui
/src/plugins/apm_oss/  @elastic/apm-ui               # @@probot-email:manu@email.com
```

cc/ @v1v @kuisathaverat

### PR DESCRIPTION
**Pull request by mdelapenya**
## What does this PR do?
It replaces the logic to parse CODEOWNERS and instead uses a dedicated YML file for it (as many of the existing probots). This new YML file's name is `.github/codeowners-labeller.yml`, and its structure is:
- yaml keys are the labels, and each label could contain an array of paths.
- for negating paths, please use an empty key (` - ""`):

```yaml
- "label1":
    - "*"
- "":
    - "/excluded/path"
```

## Why is it important?
The CODEOWENERS file, although valuable, demonstrated not be a good place to add metadata (in the form of an in-line comment with the labels for a particular path-owner).

For that reason we decided going with a separate, specific file, which simplifies the parser and, IMHO most important, opens the probot to anybody.

cc/ @v1v @kuisathaverat

## Testing
To run unit tests, execute `npm install && npm run test` from the root directory.

To verify the probot in runtime, please refer to the `How to test this` section in a previous PR: https://github.com/mdelapenya/probot-codeowners-labeller/pull/4

## Follow-ups
- what is better, to use paths or labels as YML keys? 🤔 

a) paths as keys:
```yaml
---
- "*":
  - "Good First Issue"
- "index.js":
  - "Probot"
- "lib/codeowners.js":
  - "Codeowners"
  - "NodeJS"
- "lib/owner-labeller.js":
  - "Labeller"
  - "NodeJS"
```

b) labels as keys
```yaml
---
- "Good First Issue":
  - "*"
- "Probot":
  - "index.js"
- "Codeowners":
  - "lib/codeowners.js"
- "NodeJS":
  - "lib/codeowners.js"
  - "lib/owner-labeller.js"
- "Labeller":
  - "lib/owner-labeller.js"
```

## Related issues
Closes #5
Superseeds #6

### CODE REVIEW

**Review by v1v — COMMENTED** [review]
Looks good to me. Just a comment regarding the file format, as far as I see it's `glob` based, might be interesting to see whether that's enough or it might be worth to use `regex` format instead. 

IIUC, an update in the README.md might be required for this PR too

**Comment by mdelapenya** [comment]
The minimatch library does exactly that: https://www.npmjs.com/package/minimatch

**Comment by mdelapenya** [comment]
Thinking about it again, @v1v do you mean setting up regexs as keys in the YAML file?

**Comment by v1v** [comment]
> Thinking about it again, @v1v do you mean setting up regex as keys in the YAML file?

Either way, the supported file matching is glob based, my question was whether it makes sense to use regex patterns instead of glob. Besides, IIRC, the glob format does not allow negation, in case it's a required use case.

I'd rather prefer `labels as keys`, but no strong opinion

**Comment by mdelapenya** [comment]
I'd go with the simplest path: using globs instead of regex, at least by now. We can elaborate a proposal for using regex for a follow-up conversation.

About negations, this is a use case to be considered too. Could this be a valid one?

```cucumber
When the path "A/to/B.file" is not modified
Then apply the label "LABEL"
```

And for the key/value discussion, I agree: having labels as keys sounds more reasonable for me, there could be potentially less labels than possible affected files

**Comment by kuisathaverat** [comment]
about `labels` as key vs `path` as key, we have to evaluate what option makes us repeat fewer definitions:

* `labels` usually are short texts vs `path` are long texts
* a `Path` can have several `labels`, also,  a `label` also can be applied to several `paths`, but which one has this case more often
* Which one is easy to maintain in terms of readability, Is more easy to find a path and check if the label is there or the other way around?
* What number of `labels` we can have on a repo? vs What number of `paths` we can have in the same repo? 

for me `labels` are something common to all repos (more or less), so it sounds like a good chose to select them as keys, but if we will repeat tons of text (`paths`) because we choose `labels` maybe I would choose `paths` that it is a local key to the repo.

**Comment by v1v** [comment]
> About negations, this is a use case to be considered too. Could this be a valid one?

I meant, a more complex subset of a glob. 

```cucumber
Given the path "A/to" with dozens of files
When the path "A/to/*.file" is modified 
   And the path "A/to/foo.file" is excluded
Then apply the label "LABEL"
```

Not sure if that's a valid use case though but glob might not be as flexible as the regex pattern. No strong opinion just some thoughts

**Comment by mdelapenya** [comment]
Yeah, I see your point now. I believe it's very complicated to implement that in YAML.

If we have an entry (label) for all paths under `A/to/*.file`, then do NOT apply that label if the modified file is `A/to/foo.file`

I can picture a situation where overrides apply: create an empty key at the end of the YAML file with all exclusions, and if the label is empty, the probot won't add a label.

```yaml
- "label"
  - "A/to/*.file"
- ""
  - "A/to/foo.file"
```

Wdyt?

**Comment by v1v** [comment]
I'd say if that particular use case is required then the regex is likely the best option.

Maybe the below format could be the one to support glob and regex in the future:

```
- label
   - <pattern>
       - type: (glob|regex)  glob is the default value
```


```
- "label1"
  - "A/to/[0-9]+.t*t"
      - type: regex
- "label2"              # to include all the files in A/to/* but exclude A/to/foo*
  - "A/to/(?!foo).*"
      - type: regex
- "label3"
  - "**/*.foo"
- "label4"
  - "**/*.bar"
      - type: glob
```

The below test strings might help to understand that particular use case:
image


For the time being, I'd say we don't need to think about that particular use case now

**Comment by mdelapenya** [comment]
I like the proposal, because for regex it will be simply using the defined regex.

I'd keep this initiative in a separate issue, so we can work on it for a future/next release.

BTW, I guess the dot `.` after "label2" is a typo, right?
```
- "label2" .   # to include all the files in A/to/* but exclude A/to/foo*
```

File: README.md [hunk 1]
```diff
@@ -1,12 +1,38 @@
...
-> a GitHub App built with [Probot](https://github.com/probot/probot) that adds labels in Pull Requests based on contents of the CODEOWNERS.
+> a GitHub App built with [Probot](https://github.com/probot/probot) that adds labels in Pull Requests based on contents of the `codeowners-labeller.yml` file.
 
 ## Usage
 
 1. **[Install the app](https://github.com/apps/elastic-codeowners-labeller)**.
-2. Create a `CODEOWNERS` file in your repository, under the `.github` directory.
-3. Wait for new Pull Requests to be opened or synchronised.
+1. Create a `codeowners-labeller.yml` file in your repository, under the `.github` directory.
+1. Use your own labels as the keys in the YAML file.
```
Comment (cachedout) [inline README.md:9]: Should this list be sequential? (i.e. 1, 2, 3)

**Comment by v1v** [comment]
> BTW, I guess the dot . after "label2" is a typo, right?

Indeed, good catch :)

File: README.md [hunk 2]
```diff
@@ -1,12 +1,38 @@
...
+1. Use an array for paths to be added to a label.
```
Comment (v1v) [inline README.md]: I'd say something about the format of the paths, supported `GLOB` format?

```suggestion
1. Use an array for paths to be added to a label. Pathname Format: glob
```

File: README.md (hunk 1 above)
Comment (mdelapenya) [reply README.md:9]: The cool thing about markdown is that it preserves the order, so we do not have to change different lines when reordering an item, or adding/deleting one. I discovered this a few days ago :)

Comment (cachedout) [reply README.md:9]: Ah, neat! I didn't know that. :)

File: README.md [hunk 3]
```diff
@@ -1,12 +1,38 @@
...
+1. Use your own labels as the keys in the YAML file.
+1. Use an array for paths to be added to a label. Pathname Format: glob
+1. Use the empty label (`- "":`) for exclusions. Paths here won't receive a label unless precedent labels apply.
+1. Wait for new Pull Requests to be opened or synchronised.
 
-## Important considerations
-The labels are based on Elastic criteria, so please update your [`labels.js`](./lib/labels.js) file in consequence.
+An example `codeowners-labeller.yml` file is shown here:
+
+```yaml
```
Comment (kuisathaverat) [inline README.md:16]: I'd add some examples like `folder/**/*.md` to show an example of all supported glob paths

File: lib/codeowners.js [hunk 4]
```diff
@@ -1,65 +1,89 @@
...
 
-  contents.split('\n').forEach((line) => {
-    const trimmedLine = line.trim();
+  const pathsForLabels = yaml.safeLoad(Buffer.from(contents).toString());
 
-    // discard comments
-    if (trimmedLine.startsWith('#')) {
-      return;
+  for (const label in pathsForLabels) {
+    if (!pathsForLabels.hasOwnProperty(label)) {
```
Comment (kuisathaverat) [inline lib/codeowners.js:10]: ~I did not find what it is for the `own property`~ forget it

File: lib/codeowners.js [hunk 5]
```diff
@@ -1,65 +1,89 @@
...
-        lineGlobs.push(token);
-      }
-    });
+    const pathsForLabel = pathsForLabels[label];
 
-    lineGlobs.forEach((glob) => {
-      globOwners[glob] = lineOwners;
-    });
-  });
+    const labels = Object.keys(pathsForLabel);
```
Comment (kuisathaverat) [inline lib/codeowners.js:16]: ~pathsForLabel is a list, isn't it? I've missed something~ I see it now

**Review by kuisathaverat — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

File: lib/codeowners.js (hunk 5 above)
Comment (mdelapenya) [reply lib/codeowners.js:16]: Loving NodeJS????

**Pull request merged**